File: app/src/main/java/com/platdmit/simplecloudmanager/app/vm/factory/MainActivityViewModelFactory.java
package com.platdmit.simplecloudmanager.app.vm.factory;

import com.platdmit.simplecloudmanager.domain.helpers.ActualApiKeyService;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class MainActivityViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    private ActualApiKeyService mActualApiKeyService;

    public MainActivityViewModelFactory(ActualApiKeyService actualApiKeyService) {
        super();
        mActualApiKeyService = actualApiKeyService;
    }

    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        try {
            return (T)modelClass.getConstructor(ActualApiKeyService.class).newInstance(mActualApiKeyService);
        } catch (Exception e){
            try {
                return modelClass.newInstance();
            } catch (InstantiationException | IllegalAccessException g) {
                throw new RuntimeException("Cannot create an instance of " + modelClass, g);
            }
        }
    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/vm/factory/MainActivityViewModelFactory.kt
package com.platdmit.simplecloudmanager.vm.factory

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import com.platdmit.data.helpers.ActualApiKeyService
import com.platdmit.domain.helpers.ActualApiKeyServiceManager

class MainActivityViewModelFactory(private val mActualApiKeyServiceManager: ActualApiKeyServiceManager) : NewInstanceFactory() {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return try {
            modelClass.getConstructor(ActualApiKeyServiceManager::class.java).newInstance(mActualApiKeyServiceManager)
        } catch (e: Exception) {
            try {
                modelClass.newInstance()
            } catch (g: InstantiationException) {
                throw RuntimeException("Cannot create an instance of $modelClass", g)
            } catch (g: IllegalAccessException) {
                throw RuntimeException("Cannot create an instance of $modelClass", g)
            }
        }
    }

}